### DISCUSSION TITLE
brew audit and msgpack

### DISCUSSION THREAD

**Discussion by adavidzh** [Homebrew/discussions#3767] [Everyday usage]
I was adding a new cask and going through the [testing and auditing steps](https://docs.brew.sh/Adding-Software-to-Homebrew#testing-and-auditing-the-cask). While `install` and `uninstall` work without issue on a M1 Mac, `brew audit` was very unhappy (abridged output):
```shell
$ brew audit --debug --cask --new diffusionbee
Fetching gem metadata from https://rubygems.org/.......
[...]
Fetching msgpack 1.6.0
[...]
Installing msgpack 1.6.0 with native extensions
Gem::Ext::BuildError: ERROR: Failed to build gem native extension.

    current directory: /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/msgpack-1.6.0/ext/msgpack
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby -I /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0 -r
./siteconf20221010-44732-h1rijv.rb extconf.rb
checking for ruby/st.h... *** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
	--with-opt-dir
	--without-opt-dir
	--with-opt-include
	--without-opt-include=${opt-dir}/include
	--with-opt-lib
	--without-opt-lib=${opt-dir}/lib
	--with-make-prog
	--without-make-prog
	--srcdir=.
	--curdir
	--ruby=/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/$(RUBY_BASE_NAME)
/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:467:in `try_do': The compiler failed to generate an executable file. (RuntimeError)
You have to install development tools first.
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:585:in `block in try_compile'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:534:in `with_werror'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:585:in `try_compile'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:1109:in `block in have_header'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:959:in `block in checking_for'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:361:in `block (2 levels) in postpone'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:331:in `open'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:361:in `block in postpone'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:331:in `open'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:357:in `postpone'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:958:in `checking_for'
	from /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/lib/ruby/2.6.0/mkmf.rb:1108:in `have_header'
	from extconf.rb:3:in `<main>'

To see why this extension failed to compile, please check the mkmf.log which can be found here:

  /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/extensions/universal-darwin-21/2.6.0/msgpack-1.6.0/mkmf.log

extconf failed, exit code 1

Gem files will remain installed in /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/gems/msgpack-1.6.0 for inspection.
Results logged to /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/extensions/universal-darwin-21/2.6.0/msgpack-1.6.0/gem_make.out

An error occurred while installing msgpack (1.6.0), and Bundler cannot continue.
Make sure that `gem install msgpack -v '1.6.0' --source 'https://rubygems.org/'` succeeds before bundling.

In Gemfile:
  bootsnap was resolved to 1.13.0, which depends on
    msgpack
Error: failed to run `/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/bundle install`!
Error: Kernel.exit
/opt/homebrew/Library/Homebrew/utils.rb:164:in `exit'
/opt/homebrew/Library/Homebrew/utils.rb:164:in `odie'
/opt/homebrew/Library/Homebrew/utils/gems.rb:46:in `odie_if_defined'
/opt/homebrew/Library/Homebrew/utils/gems.rb:165:in `install_bundler_gems!'
/opt/homebrew/Library/Homebrew/style.rb:86:in `run_rubocop'
/opt/homebrew/Library/Homebrew/style.rb:54:in `check_style_impl'
/opt/homebrew/Library/Homebrew/style.rb:36:in `check_style_json'
/opt/homebrew/Library/Homebrew/dev-cmd/audit.rb:185:in `audit'
/opt/homebrew/Library/Homebrew/brew.rb:95:in `<main>'
```

The mentioned `mkmf.log` said that
```shell
cat /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/extensions/universal-darwin-21/2.6.0/msgpack-1.6.0/mkmf.log
───────┬────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
       │ File: /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0/extensions/universal-darwin-21/2.6.0/msgpack-1.6.0/mkmf.log
───────┼────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
   1   │ "xcrun clang -o conftest -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/include/ruby-2.6.0/uni
       │ versal-darwin21 -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/include/ruby-2.6.0/ruby/backwar
       │ d -I/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/include/ruby-2.6.0 -I. -D_XOPEN_SOURCE -D_DAR
       │ WIN_C_SOURCE -D_DARWIN_UNLIMITED_SELECT -D_REENTRANT    -g -Os -pipe -DHAVE_GCC_ATOMIC_BUILTINS -DUSE_FFI_CLOSURE_ALLOC conftest.c  -L. -L/System/Library/Frame
       │ works/Ruby.framework/Versions/2.6/usr/lib -L. -L/AppleInternal/Library/BuildRoots/a0876c02-1788-11ed-b9c4-96898e02b808/Applications/Xcode.app/Contents/Develope
       │ r/Platforms/MacOSX.platform/Developer/SDKs/MacOSX12.6.Internal.sdk/usr/local/lib     -lruby.2.6   "
   2   │ In file included from conftest.c:1:
   3   │ In file included from /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/include/ruby-2.6.0/ruby.h:3
       │ 3:
   4   │ /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/include/ruby-2.6.0/ruby/ruby.h:24:10: fatal error
       │ : 'ruby/config.h' file not found
   5   │ #include "ruby/config.h"
   6   │          ^~~~~~~~~~~~~~~
   7   │ /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/include/ruby-2.6.0/ruby/ruby.h:24:10: note: did n
       │ ot find header 'config.h' in framework 'ruby' (loaded from '/Library/Developer/CommandLineTools/SDKs/MacOSX12.sdk/System/Library/Frameworks')
   8   │ 1 error generated.
   9   │ checked program was:
  10   │ /* begin */
  11   │ 1: #include "ruby.h"
  12   │ 2:
  13   │ 3: int main(int argc, char **argv)
  14   │ 4: {
  15   │ 5:   return 0;
  16   │ 6: }
  17   │ /* end */
  18   │
───────┴────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
```

I could not find good ways around the failure to build `msgpack 1.6.0` and [other threads](https://github.com/Homebrew/discussions/discussions/716) are not very informative.

So I tried the stupid solution of installing `msgpack` myself:
```shell
/usr/bin/gem install -i /opt/homebrew/Library/Homebrew/vendor/bundle/ruby/2.6.0  msgpack
Building native extensions. This could take a while...
Successfully installed msgpack-1.6.0
Parsing documentation for msgpack-1.6.0
Installing ri documentation for msgpack-1.6.0
Done installing documentation for msgpack after 0 seconds
1 gem installed
```

And that worked! I have no idea why this would work and what `brew audit` was trying to do did not work. Perhaps environment variables?

So, basically I am leaving this here in case someone else stumbles upon this problem (that has such a simple solution).

**Comment by afragen** [comment]
I ran into the same error and found it resolved after running `brew install-bundler-gems`

**Reply by xxyzz** [reply]
I run the command but still get the same error.
